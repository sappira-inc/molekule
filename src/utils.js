import { get, kebabCase } from 'lodash';
import styled from 'styled-components';

export const themeGet = (theme, lookup, fallback) => get(theme, lookup, fallback);

export const getComponentVariant = (theme, componentName, variant) => {
  const config = themeGet(
    theme,
    `components.${componentName}.variants.${variant}`,
    theme.variants[componentName][variant]
  );
  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentStyle = componentName => ({ theme }) => themeGet(theme, `components.${componentName}.style`);

export const getComponentClassName = ({ className, theme: { classPrefix }, variant }, name) =>
  `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

export const createComponent = ({ name, tag = 'div', as, style, props: defaultProps = () => ({}) }) => {
  const component = as ? styled(as) : styled[tag];

  return component.attrs(props => ({
    ...defaultProps(props),
    ...props,
    className: getComponentClassName(props, kebabCase(name)),
  }))`
    ${style}
    ${getComponentStyle(name)}
    ${({ styles = {} }) => styles[name] || {}}
  `;
};
