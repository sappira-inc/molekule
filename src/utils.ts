import kebabCase from 'lodash/kebabCase';
import get from 'lodash/get';
import styled, { CSSProp } from 'styled-components';

export const themeGet = (lookup: any, fallback?: any) => ({ theme }: any = {}) => get(theme, lookup, fallback);

export const getComponentVariant = (theme: any, componentName: string, variant: string) => {
  const config = themeGet(`variants.${componentName}.${variant}`)({
    theme,
  });

  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentSize = (theme: any, componentName: string, size: string) =>
  themeGet(`sizes.${componentName}.${size}`, {})({ theme });

export const getComponentStyle = (componentName: string) => themeGet(`styles.${componentName}`, {});

export const getVariantStyles = (componentName: string, variant: string) =>
  themeGet(`variants.${componentName}.${variant}.style`, {});

interface GetComponentClassNameProps {
  className: string;
  theme: {
    classPrefix: string;
  };
  variant: string;
}

const getComponentClassName = (
  { className, theme: { classPrefix }, variant }: GetComponentClassNameProps,
  name: string
) => `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

interface CreateComponentProps<T> {
  name: string;
  tag?: keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>;
  as?: React.ComponentType<any>;
  style?: ((props: any) => CSSProp) | CSSProp;
  props?: (props: T) => any;
}

export const createComponent = <
  T extends object,
  O extends keyof JSX.IntrinsicElements | React.ComponentType<any> = 'div'
>({
  name,
  tag = 'div',
  as,
  style,
  props: getBaseProps = () => ({}),
}: CreateComponentProps<T>) => {
  const component = styled<O>((as || tag) as any);

  return component.attrs<T>((props: any) => {
    const baseProps = getBaseProps(props);
    const finalProps = {
      ...baseProps,
      ...props,
      className: baseProps.className || props.className,
    };

    return {
      ...finalProps,
      className: getComponentClassName(finalProps, kebabCase(name)),
    };
  })<T>`
  ${style}
  ${getComponentStyle(name)}
  ${(p: any) => getVariantStyles(name, p.variant)}
  ${({ styles = {} }) => styles[name] || {}}
`;
};

// eslint-disable-next-line
const focusableFilter = (node: any) =>
  node.tabIndex >= 0 && !node.hasAttribute('disabled') ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP;

type FilterFn = (node?: any) => any;

const createTreeWalker = (root: any, currentNode: any, filterFn: FilterFn = () => NodeFilter.FILTER_ACCEPT) => {
  const treeWalker = document.createTreeWalker(
    root,
    NodeFilter.SHOW_ELEMENT,
    {
      acceptNode: filterFn,
    },
    false
  );
  treeWalker.currentNode = currentNode;
  return treeWalker;
};

export const findPreviousFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.previousNode() || treeWalker.currentNode === root) {
    return treeWalker.lastChild();
  }

  return treeWalker.currentNode;
};

export const findNextFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.nextNode() || treeWalker.currentNode === root) {
    treeWalker.currentNode = root;
    return treeWalker.firstChild();
  }

  return treeWalker.currentNode;
};

export const getNextCursorPosition = (cursorPos: any, oldValue: any, newValue: any) => {
  const cursorDiff = newValue.length - oldValue.length;
  let nextPosition = cursorPos;

  if (cursorDiff > 1 || cursorDiff < -1) {
    nextPosition += cursorDiff < 0 ? cursorDiff + 1 : cursorDiff - 1;
  }

  return Math.max(nextPosition, 0);
};
