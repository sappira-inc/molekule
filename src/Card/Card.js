import React from 'react';
import PropTypes from 'prop-types';
import { css } from 'styled-components';
import Box from '../Box';
import { themeGet, createComponent } from '../utils';

const StyledCard = createComponent({
  name: 'Card',
  as: Box,
  style: ({ shadow, theme }) => css`
    background: white;
    box-shadow: ${shadow ? themeGet('shadow.soft', 'none') : 'none'};
    border-radius: ${themeGet('radius', 0)}px;
    overflow: hidden;
    border: 1px solid ${theme.colors.greyLight};
  `,
});

/** Cards provide a flexible way to encapsulate content with multiple variants and options. */
const Card = React.forwardRef((props, ref) => <StyledCard ref={ref} {...props} />);

Card.propTypes = {
  shadow: PropTypes.bool,
};

Card.defaultProps = {
  shadow: false,
};

Card.Header = createComponent({
  name: 'CardHeader',
  as: Box,
  style: css`
    padding: 1rem 1rem 8px;
    font-weight: 700;
  `,
});

Card.Body = createComponent({
  name: 'CardBody',
  as: Box,
  style: () => css`
    padding: 8px 1rem;
  `,
});

Card.Footer = createComponent({
  name: 'CardFooter',
  as: Box,
  style: css`
    padding: 8px 1rem 1rem;
  `,
});

export default Card;
