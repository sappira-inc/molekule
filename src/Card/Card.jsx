import React from 'react';
import PropTypes from 'prop-types';
import { css } from 'styled-components';
import { space, borderRadius, themeGet } from 'styled-system';
import { createComponent } from '../utils';

const StyledCard = createComponent({
  name: 'Card',
  style: ({ shadow }) => css`
    background: white;
    box-shadow: ${shadow
      ? themeGet('shadow', '0 4px 10px rgba(0, 29, 54, 0.04), 0 -2px 6px rgba(0, 29, 54, 0.025);')
      : 'none'};

    ${borderRadius};
    ${space};
  `,
});

const Card = props => <StyledCard {...props} />;

Card.propTypes = {
  shadow: PropTypes.bool,
};

Card.defaultProps = {
  shadow: true,
};

Card.Header = createComponent({
  name: 'CardHeader',
  style: () => css`
    padding: 1rem;
    border-bottom: 1px solid ${themeGet('colors.grayLight')};

    ${space};
  `,
});

Card.Body = createComponent({
  name: 'CardBody',
  style: () => css`
    padding: 1rem;

    ${space};
  `,
});

Card.Footer = createComponent({
  name: 'CardFooter',
  style: () => css`
    padding: 1rem;
    border-top: 1px solid ${themeGet('colors.grayLight')};

    ${space};
  `,
});

export default Card;
