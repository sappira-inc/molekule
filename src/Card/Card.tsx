import React from 'react';
import { css } from 'styled-components';
import Box, { BoxProps } from '../Box';
import { themeGet, createComponent } from '../utils';

export interface CardProps {
  shadow?: boolean;
  children?: React.ReactNode;
}

const StyledCard = createComponent<CardProps & BoxProps>({
  name: 'Card',
  as: Box,
  style: ({ shadow, theme }) => css`
    background: white;
    box-shadow: ${shadow ? themeGet('shadow.soft', 'none') : 'none'};
    border-radius: ${themeGet('radius', 0)}px;
    overflow: hidden;
    border: 1px solid ${theme.colors.greyLight};
  `,
});

export interface CardStaticMembers {
  Footer: React.ComponentType<BoxProps>;
  Body: React.ComponentType<BoxProps>;
  Header: React.ComponentType<BoxProps>;
}

/** Cards provide a flexible way to encapsulate content with multiple variants and options. */
const Card = React.forwardRef<HTMLDivElement, CardProps & BoxProps>((props, ref) => (
  <StyledCard ref={ref} {...props} />
)) as React.ForwardRefExoticComponent<CardProps & BoxProps & React.RefAttributes<HTMLDivElement>> & CardStaticMembers;

Card.defaultProps = {
  shadow: false,
};

Card.Footer = createComponent<BoxProps>({
  name: 'CardFooter',
  as: Box,
  style: css`
    padding: 16px;
  `,
});

const cardFooterClass = Card.Footer.toString();

Card.Body = createComponent<BoxProps>({
  name: 'CardBody',
  as: Box,
  style: () => css`
    padding: 16px;

    & + ${cardFooterClass} {
      padding-top: 0px;
    }
  `,
});

const cardBodyClass = Card.Body.toString();

Card.Header = createComponent<BoxProps>({
  name: 'CardHeader',
  as: Box,
  style: css`
    padding: 16px;
    font-weight: 700;

    & + ${cardBodyClass} {
      padding-top: 0px;
    }
  `,
});

export default Card;
