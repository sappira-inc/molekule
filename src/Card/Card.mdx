---
menu: Components
name: Card
---

import { Playground, PropsTable } from 'docz'
import Card from './Card'

# Card

Cards provide a flexible way to encapsulate content with multiple variants and options.

## Props

<PropsTable of={Card} />

## Example

<Playground>
  <Card>
    <Card.Header>Header</Card.Header>
    <Card.Body>
      A card is a flexible and extensible content container. It includes options for headers and footers, a wide variety of content, contextual background colors, and powerful display options.
    </Card.Body>
    <Card.Footer>Footer</Card.Footer>
  </Card>
</Playground>
