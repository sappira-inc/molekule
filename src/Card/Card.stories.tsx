import React from 'react';
import Card from './Card';

export default {
  title: 'Components|Card',
  component: Card,
};

export const Basic = () => (
  <Card>
    <Card.Header>Header</Card.Header>
    <Card.Body>
      A card is a flexible and extensible content container. It includes options for headers and footers, a wide variety
      of content, contextual background colors, and powerful display options.
    </Card.Body>
    <Card.Footer>Footer</Card.Footer>
  </Card>
);

export const BodyOnly = () => (
  <Card>
    <Card.Body>
      A card is a flexible and extensible content container. It includes options for headers and footers, a wide variety
      of content, contextual background colors, and powerful display options.
    </Card.Body>
  </Card>
);
