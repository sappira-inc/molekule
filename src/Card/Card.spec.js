import React from 'react';
import { renderWithTheme } from '../../test/utils';
import Card from './Card';

test('Card', () => {
  const { asFragment } = renderWithTheme(
    <Card>
      <Card.Header>Header</Card.Header>
      <Card.Body>Body</Card.Body>
      <Card.Footer>Footer</Card.Footer>
    </Card>
  );

  expect(asFragment()).toMatchSnapshot();
});
