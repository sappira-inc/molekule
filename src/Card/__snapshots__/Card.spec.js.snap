// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card 1`] = `
.c0 {
  background: white;
  box-shadow: 0 3px 6px hsla(0,0%,60%,.1),0 3px 6px hsla(0,0%,60%,.15),0 -1px 2px hsla(0,0%,60%,.02);
  border-radius: 4px;
  overflow: hidden;
}

.c1 {
  padding: 1rem;
  border-bottom: 1px solid #DEE0E4;
}

.c2 {
  padding: 1rem;
}

.c3 {
  padding: 1rem;
  border-top: 1px solid #DEE0E4;
}

<div
  className="re-box re-card c0"
>
  <div
    className="re-box re-card-header c1"
  >
    Header
  </div>
  <div
    className="re-box re-card-body c2"
  >
    Body
  </div>
  <div
    className="re-box re-card-footer c3"
  >
    Footer
  </div>
</div>
`;
